bugrap/src/main/java/org/vaadin/bugrap/ui/ReportDetailView.java

```diff
@@ -119,55 +116,61 @@ private void saveComment() {
 
 	@Override
 	public void uploadStarted(StartedEvent event) {
-		pnlAttachments.setVisible(true);
+		initiateUploadTracker(event);
+		addNewUpload();
+		numOfOngoingUploads++;
+	}
+
+	private void addNewUpload() {
+		HorizontalLayout layout = (HorizontalLayout) btnUpload.getParent();
+		Upload newUpload = new Upload();
+		newUpload.setButtonCaption("Attachments...");
+		newUpload.addStartedListener(this);
+		newUpload.addFinishedListener(this);
+		layout.addComponent(newUpload, layout.getComponentIndex(btnUpload));
+		btnUpload.setStyleName("removed-upload", true);
+		btnUpload = newUpload;
+	}
+
+	private void initiateUploadTracker(StartedEvent event) {
+		UploadTracker uploadTracker = new UploadTracker((Upload) event.getSource(), event.getFilename());
+		uploadTracker.setUploadedFileListener(this);
 		HorizontalLayout layout = (HorizontalLayout) pnlAttachments.getContent();
-		ProgressBar progress = new ProgressBar();
-		progress.setCaption(event.getFilename());
-		uploadInProgress = progress;
-		layout.addComponent(progress);
-		btnDone.setEnabled(false);//disable done button until upload completes
+		layout.addComponent(uploadTracker);
+	}
+
+	private void closeWindow() {
+		Window window = (Window) getParent();
+		window.close();
 	}
 
 	@Override
-	public void uploadSucceeded(SucceededEvent event) {
-		HorizontalLayout layout = (HorizontalLayout) pnlAttachments.getContent();
-		layout.removeComponent(uploadInProgress);
-		uploadInProgress = null;
-		FileResource file = new FileResource(new File(ReportsModel.FILEUPLOAD_PATH + event.getFilename()));
-		try {
-			model.attachFile(event.getFilename(), event.getMIMEType(), file.getStream());
-		} catch (IOException e) {
-			Notification.show("Error occurred while uploading the file", e.getMessage(), Type.ERROR_MESSAGE);
-		}
-		FileDownloader downloader = new FileDownloader(file);
-		Link link = new Link();
-		link.setCaption(event.getFilename());
-		link.setStyleName("tiny");
-		downloader.extend(link);
-		layout.addComponent(link);
-		commentsUpdated();
+	public void componentEvent(Event event) {
```

Comment:
Is componentEvent method being used?

Reply:
yep, since ReportDetailView implements Component.Listener, it is used whenever a component added or removed from the container